knitr/hmm-example/hmm-example.Rmd

```diff
@@ -0,0 +1,287 @@
+    log_omega[1, n] = normal_lpdf(y[n] | mu[1], sigma);
+    log_omega[2, n] = normal_lpdf(y[n] | mu[2], sigma);
+    log_omega[3, n] = normal_lpdf(y[n] | mu[3], sigma);
+  }
+}
+```
+
+With all that in place, the only thing left to do is add priors and increment
+the log density:
+```{stan, output.var = "", eval = FALSE}
+model {
+  mu ~ normal(0, 1);
+  sigma ~ normal(0, 1);
+
+  target += hmm_marginal(log_omega, Gamma, rho);
+}
+```
+
+The complete model is available on Github: [hmm-example.stan](https://github.com/stan-dev/example-models/tree/master/knitr/hmm-example/hmm-example.stan).
+
+```{r echo = TRUE, results = FALSE, message = FALSE}
+model = cmdstan_model("hmm-example.stan")
+fit = model$sample(data = list(N = N, y = y), parallel_chains = 4)
+```
+
+The estimated group means match the known ones:
+```{r}
+fit$summary("mu")
+```
+The estimated initial conditions are not much more informative than
+the prior, but it is there:
+```{r}
+fit$summary("rho")
+```
+
+The transition probabilities from state 1 can be extracted:
+```{r}
+fit$summary("t1")
+```
+
+Similarly for state 2:
+```{r}
+fit$summary("t2")
+```
+
+And state 3:
+```{r}
+fit$summary("t3")
+```
+
+### State Probabilities
+
+Even though the hidden states are integrated out, the distribution
+of hidden states at each time point can be computed with the function
+`hmm_hidden_state_prob`:
+
+```{stan, output.var = "", eval = FALSE}
+generated quantities {
+  matrix[3, N] hidden_probs = hmm_hidden_state_prob(log_omega, Gamma, rho);
+}
+```
+
+These can be plotted:
+
+```{r}
+hidden_probs_df = fit$draws() %>%
+  as_draws_df %>%
+  select(starts_with("hidden_probs")) %>%
+  pivot_longer(everything(),
+               names_to = c("state", "n"),
+               names_transform = list(k = as.integer, n = as.integer),
+               names_pattern = "hidden_probs\\[([0-9]*),([0-9]*)\\]",
+               values_to = "hidden_probs")
+
+hidden_probs_df %>%
+  group_by(state, n) %>%
+  summarize(qh = quantile(hidden_probs, 0.8),
+            m = median(hidden_probs),
+            ql = quantile(hidden_probs, 0.2)) %>%
+  ungroup() %>%
+  ggplot() +
+  geom_errorbar(aes(n, ymin = ql, ymax = qh, width = 0.0), alpha = 0.5) +
+  geom_point(aes(n, m)) +
+  facet_grid(state ~ ., labeller = "label_both") +
+  ggtitle("Ribbon is 60% posterior interval, point is median") +
+  ylab("Probability of being in state") +
+  xlab("Time (n)")
+```
+
+If it is more convenient to work with draws of the hidden states at each time
+point (instead of the probabilities provided by `hmm_hidden_state_prob`), these
+can be generated with `hmm_latent_rng`:
+
+```{stan, output.var = "", eval = FALSE}
+generated quantities {
+  int[N] y_sim = hmm_latent_rng(log_omega, Gamma, rho)
+}
+```
+
+Note, these are the probabilities and draws of the hidden state that was
+integrated out. This is different than sampling new HMM realizations using
+posterior draws of the initial condition and the transition matrix.
```

Comment:
"Note further that the probabilities are the _marginal_ probabilities, meaning they cannot be directly used to jointly sample hidden states. However, the posterior draws generated by `hmm_latent_rng` account for the correlation between hidden states."